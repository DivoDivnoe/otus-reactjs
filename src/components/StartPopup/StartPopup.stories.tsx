import React from 'react';
import { ComponentStory, ComponentMeta } from '@storybook/react';

import { StartPopup } from './';
import { Provider } from 'react-redux';
import { createStore } from 'redux';
import reducer from '@/reducer';

export default {
  title: 'Example/StartPopup',
  component: StartPopup,
  argTypes: {
    signin: { action: 'sign' },
    setName: { action: 'setName' },
    resetName: { action: 'resetName' },
  },
} as ComponentMeta<typeof StartPopup>;

const store = createStore(reducer);

const Template: ComponentStory<typeof StartPopup> = (args) => (
  <Provider store={store}>
    <StartPopup {...args} />
  </Provider>
);

export const MainStartPopup = Template.bind({});
MainStartPopup.args = {
  currentName: 'Andrey',
};
