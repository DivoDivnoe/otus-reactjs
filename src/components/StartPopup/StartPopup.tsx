import React, { Component, ReactNode } from 'react';
import styled from '@emotion/styled';
import Button from '@/components/Button/Button';

export interface PopupProps {
  isVisible: boolean;
  submitHandler: (name: string) => void;
}

export interface PopupState {
  name: string;
}

const PopupWrapper = styled.div`
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  background-color: rgba(8, 0, 18, 0.85);
`;

const PopupItem = styled.div`
  position: absolute;
  left: 50%;
  top: 50%;
  padding: 30px;
  border-radius: 7px;
  background-color: #021a60;
  transform: translate(-50%, -50%);
`;

const FormItem = styled.form`
  display: flex;
  flex-direction: column;
  width: 240px;
`;

const InputItem = styled.input`
  padding: 6px 10px 4px 10px;
  margin-bottom: 20px;
  font-size: 16px;
  line-height: 19px;
  color: #ffffff;
  border: 2px solid #058aba;
  border-radius: 3px;
  background-color: transparent;
  box-shadow: inset 1px 1px 3px rgb(0 0 0 / 30%), 1px 1px 0 rgb(0 0 0 / 10%);
`;

class StartPopup extends Component<PopupProps, PopupState> {
  state = {
    name: '',
  };

  render(): ReactNode {
    const { isVisible } = this.props;

    return (
      isVisible && (
        <PopupWrapper>
          <PopupItem>
            <FormItem data-testid='start-form' onSubmit={this._onSubmit}>
              <InputItem
                data-testid='name-input'
                value={this.state.name}
                placeholder={'Enter Your Name:'}
                onChange={this._onChange}
                required
              />
              <Button>Start</Button>
            </FormItem>
          </PopupItem>
        </PopupWrapper>
      )
    );
  }

  _onSubmit = (evt: React.FormEvent): void => {
    evt.preventDefault();

    this.props.submitHandler(this.state.name);
  };

  _onChange = (evt: React.ChangeEvent<HTMLInputElement>): void => {
    this.setState({ name: evt.target.value.trim() });
  };
}

export default StartPopup;
