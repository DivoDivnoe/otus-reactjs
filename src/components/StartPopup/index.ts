export { StartPopup, StartPopupProps } from './StartPopup';
