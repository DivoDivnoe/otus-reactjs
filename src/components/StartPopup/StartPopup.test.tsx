import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { StartPopup, StartPopupProps } from './StartPopup';

describe('StartPopup', () => {
  it('is rendered correctly', () => {
    const mockProps: StartPopupProps = {
      currentName: 'Andrey',
      signin: jest.fn(),
      setName: jest.fn(),
      resetName: jest.fn(),
    };

    const { asFragment } = render(<StartPopup {...mockProps} />);

    expect(screen.getByRole('button', { name: 'Start' })).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it('input has right value', () => {
    const mockProps: StartPopupProps = {
      currentName: 'Andrey',
      signin: jest.fn(),
      setName: jest.fn(),
      resetName: jest.fn(),
    };

    render(<StartPopup {...mockProps} />);
    expect(screen.getByTestId('name-input')).toHaveValue('Andrey');
  });

  it('handles change event correctly', () => {
    const mockProps: StartPopupProps = {
      currentName: '',
      signin: jest.fn(),
      setName: jest.fn(),
      resetName: jest.fn(),
    };

    render(<StartPopup {...mockProps} />);

    userEvent.type(screen.getByTestId('name-input'), 'Andrey');
    expect(mockProps.setName).toHaveBeenCalledTimes(6);
  });

  it('clears input after submit', () => {
    const mockProps: StartPopupProps = {
      currentName: '',
      signin: jest.fn(),
      setName: jest.fn(),
      resetName: jest.fn(),
    };

    render(<StartPopup {...mockProps} />);
    userEvent.type(screen.getByTestId('name-input'), 'Andrey');
    fireEvent.submit(screen.getByTestId('start-form'));
    expect(screen.getByTestId('name-input')).toHaveValue('');
  });
});
