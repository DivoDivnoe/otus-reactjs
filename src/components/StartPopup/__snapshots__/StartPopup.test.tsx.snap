// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StartPopup is rendered correctly 1`] = `
<DocumentFragment>
  <div
    class="css-1p0rybr"
    data-testid="start-popup"
  >
    <form
      class="css-qztj7d"
      data-testid="start-form"
    >
      <input
        class="css-1oh381d"
        data-testid="name-input"
        placeholder="Enter Your Name:"
        required=""
        value="Andrey"
      />
      <button
        class="css-som8ht"
      >
        Start
      </button>
    </form>
  </div>
</DocumentFragment>
`;
